module/state_synchronization/indexer/execution.go

```diff
@@ -0,0 +1,185 @@
+		return nil, errors.Wrap(err, "failed to initialize the indexer with missing latest height")
+	}
+
+	indexRange, err := NewSequentialIndexRange(first, last)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Debug().Msgf("initialized indexer with range first: %d, last: %d", first, last)
+
+	return &ExecutionState{
+		registers:  registers,
+		headers:    headers,
+		events:     events,
+		indexRange: indexRange,
+		log:        log.With().Str("component", "execution_indexer").Logger(),
+	}, nil
+}
+
+// RegisterValues retrieves register values by the register IDs at the provided block height.
+// Even if the register wasn't indexed at the provided height, returns the highest height the register was indexed at.
+// Expected errors:
+// - storage.ErrNotFound if the register by the ID was never indexed
+// - ErrIndexBoundary if the height is out of indexed height boundary
+func (i *ExecutionState) RegisterValues(IDs flow.RegisterIDs, height uint64) ([]flow.RegisterValue, error) {
+	_, err := i.indexRange.Contained(height)
+	if err != nil {
+		return nil, err
+	}
+
+	values := make([]flow.RegisterValue, len(IDs))
+
+	for j, id := range IDs {
+		value, err := i.registers.Get(id, height)
+		if err != nil {
+			return nil, err
+		}
+
+		values[j] = value
+	}
+
+	return values, nil
+}
+
+// IndexBlockData indexes all execution block data by height.
+// This method shouldn't be used concurrently.
+// Expected errors:
+// - ErrIndexValue if height was not incremented in sequence
+// - storage.ErrNotFound if the block for execution data was not found
+func (i *ExecutionState) IndexBlockData(ctx context.Context, data *execution_data.BlockExecutionDataEntity) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+	}
+
+	block, err := i.headers.ByBlockID(data.BlockID)
+	if err != nil {
+		// todo should we wrap sentinel error or should we use irrecoverable.exception as per design guidelines https://github.com/onflow/flow-go/blob/master/CodingConventions.md
+		return fmt.Errorf("could not get the block by ID %s: %w", data.BlockID, err)
+	}
+
+	i.log.Debug().Msgf("indexing new block ID %s at height %d", data.BlockID.String(), block.Height)
+
+	if _, err := i.indexRange.CanIncrease(block.Height); err != nil {
+		return err
+	}
+
+	// concurrently process indexing of block data
+	g, _ := errgroup.WithContext(ctx)
+
+	payloads := make(map[ledger.Path]*ledger.Payload)
+	events := make([]flow.Event, 0)
+
+	for _, chunk := range data.ChunkExecutionDatas {
+		if chunk.TrieUpdate != nil {
+			// we are iterating all the registers and overwrite any existing register at the same path
+			// this will make sure if we have multiple register changes only the last change will get persisted
+			// if block has two chucks:
+			// first chunk updates: { X: 1, Y: 2 }
+			// second chunk updates: { X: 2 }
+			// then we should persist only {X: 2: Y: 2}
+			for i, path := range chunk.TrieUpdate.Paths {
+				payloads[path] = chunk.TrieUpdate.Payloads[i]
```

Comment:
I want to double-check check it's okay to use `TrieUpdate.Paths` instead of `TrieUpdate.Payload.Key`